Text Annotation Leaders

SheetalNair opened the pull request:
This PR only consists of code to create leader geometry and store it in Text Annotation, the later part of creating relationships  between leader and target elements is not yet addressed here. Find details in this issue #7830 
I have introduced attachmentMode property for leaders which provides user with a set of options to select attachment point:
- Nearest : leader will attach to the point on frame closest to the target point.
![image](https://github.com/user-attachments/assets/336076a9-2d72-41e3-b1fa-84f910326c3d)

- KeyPoint: User can provide the curveIndex and fraction of the frame as an attachment point (See image below: curveIndex 0, fraction 0)
![image](https://github.com/user-attachments/assets/0c0c549b-dda2-40da-9b18-68360f87fa0b)

- TextPoint: User can attach the leader either to the first or last line of text (See image below: position:"TopLeft" indicating first line of text)
![image](https://github.com/user-attachments/assets/79514d4a-abb4-47a3-84ef-e4d233f7b8c4)


Code changes:
- `LeaderGeometry.ts` : contains functions to generate leader geometry and append it to builder
- `TextAnnotation.ts`: Introduced leaderprops inside TextAnnotation class
- `TextAnnotationGeometry.ts`: Call to appendLeadersToBuilder()
- `TextStyle.ts` : Introduced leader style props [Current decision is for leader style to be a part of TextStyle]
- `TextStyle.test.ts`: fixed failing test case after leader style introduction
- `TextDecoration.ts` : Keyins for testing leaders

claudiaareneee commented on test-apps/display-test-app/src/frontend/TextDecoration.ts:
```diff
@@ -363,6 +388,33 @@ export class TextDecorationTool extends Tool {
 
         break;
       }
+      case "leader":
+        const k = inArgs[1];
+        const value = inArgs[2];
+        if (k === "new") {
+          editor.setLeaderProps();
+        } else {
+          if (editor.leaders && editor.leaders.length > 0) {
+            const latestLeaderIndex = editor.leaders.length - 1;
+            if (k === "start") editor.setLeaderStartPoint(editor.leaders[latestLeaderIndex], Number(value));
+            else if (k === "keypoint") {
+              const curveIndex = inArgs[2];
+              const fraction = inArgs[3]
+              editor.setLeaderKeyPoint(editor.leaders[latestLeaderIndex], Number(curveIndex), Number(fraction));
+            }
+            else if (k === "nearest") editor.setLeaderNearest(editor.leaders[latestLeaderIndex]);
+            else if (k === "textpoint") {
+              const position = inArgs[2] as LeaderTextPointOptions;
+              editor.setLeaderTextPoint(editor.leaders[latestLeaderIndex], position);
+
+            }
+            else throw new Error("Expected start, keypoint, nearest, textpoint");
+          } else {
+            throw new Error("No leaders created. Use dta text leader new.");
+          }
+
+        }
+        break;
```
```suggestion
      case "leader":
        const attachmentType = inArgs[1];
        const value = inArgs[2];
        if (attachmentType === "new") {
          editor.setLeaderProps();
        } else {
          if (editor.leaders && editor.leaders.length > 0) {
            const latestLeaderIndex = editor.leaders.length - 1;
            if (attachmentType === "start") editor.setLeaderStartPoint(editor.leaders[latestLeaderIndex], Number(value));
            else if (attachmentType === "keypoint") {
              const curveIndex = inArgs[2];
              const fraction = inArgs[3]
              editor.setLeaderKeyPoint(editor.leaders[latestLeaderIndex], Number(curveIndex), Number(fraction));
            }
            else if (attachmentType === "nearest") editor.setLeaderNearest(editor.leaders[latestLeaderIndex]);
            else if (attachmentType === "textpoint") {
              const position = inArgs[2] as LeaderTextPointOptions;
              editor.setLeaderTextPoint(editor.leaders[latestLeaderIndex], position);

            }
            else throw new Error("Expected start, keypoint, nearest, textpoint");
          } else {
            throw new Error("No leaders created. Use dta text leader new.");
          }

        }
        break;
```

claudiaareneee replied:
Can you rename `k` to something more explicit. Maybe `attachmentType` or something?

My suggestion above uses `attachmentType`

SheetalNair replied:
all keys are not attachment types so I used the word 'command'

claudiaareneee commented on test-apps/display-test-app/src/frontend/TextDecoration.ts:
```diff
@@ -3,11 +3,11 @@
+import { Angle, Point3d, Vector3d, YawPitchRollAngles } from "@itwin/core-geometry";
 
 // Ignoring the spelling of the keyins. They're case insensitive, so we check against lowercase.
 // cspell:ignore superscript, subscript, widthfactor, fractionscale, fractiontype
```
```suggestion
// cspell:ignore superscript, subscript, widthfactor, fractionscale, fractiontype, textpoint
```

claudiaareneee commented on core/common/src/annotation/TextStyle.ts:
```diff
@@ -141,6 +164,22 @@ export class TextStyleSettings {
+  public readonly leaderColor: TextStyleColor;
+  /** Whether to use an elbow in the leader line. */
+  public readonly wantElbow: boolean;
+  /** Multiplier used to compute length of the elbow in the leader line.
+   * The elbowLength is computed as elbowLength * [[lineHeight]].
+  */
+  public readonly elbowLength: number;
+  /** Multiplier to compute height of the leader terminator.
+   * The terminator height is computed as terminatorHeight * [[lineHeight]].
+  */
+  public readonly terminatorHeight: number;
+  /** Multiplier to compute width of the leader terminator.
+   * The terminator width is computed as terminatorWidth * [[lineHeight]].
+  */
+  public readonly terminatorWidth: number;
```
```suggestion
  /** Multiplier used to compute length of the elbow in the leader line.
   * The elbowLength is computed in meters as elbowLength * [[lineHeight]].
  */
  public readonly elbowLength: number;
  /** Multiplier to compute height of the leader terminator.
   * The terminator height is computed in meters as terminatorHeight * [[lineHeight]].
  */
  public readonly terminatorHeight: number;
  /** Multiplier to compute width of the leader terminator.
   * The terminator width is computed in meters as terminatorWidth * [[lineHeight]].
  */
  public readonly terminatorWidth: number;
```

claudiaareneee commented on core/common/src/annotation/TextStyle.ts:
```diff
@@ -88,6 +88,29 @@ export interface TextStyleSettingsProps {
+   * Default: false
+   */
+  wantElbow?: boolean;
+  /** Multiplier used to compute length of the elbow in the leader line.
+   * The elbowLength is computed as elbowLength * [[lineHeight]].
+   * Default: 1.0
+   */
+  elbowLength?: number;
+  /** Multiplier to compute height of the leader terminator.
+   * The terminator height is computed as terminatorHeight * [[lineHeight]].
+   * Default: 1.0
+   */
+  terminatorHeight?: number;
+  /** Multiplier to compute width of the leader terminator.
+   * The terminator width is computed as terminatorWidth * [[lineHeight]].
+   * Default: 1.0
+   */
+  terminatorWidth?: number;
```
Also maybe these should be renamed with a `Factor` at the end to maintain the consistency with the other multiplier settings? Right now the sound like the explicit width/length. That might make the attributes a bit long though. They could be nested inside a `terminator` variable? I know @Alfonso-Martello is working on styles right now so maybe I should just wait for those comments, idk.

SheetalNair replied:
Appended the variable name with 'Factor', we had a discussion some time back about not nesting properties in style.

claudiaareneee commented on core/backend/src/annotations/LeaderGeometry.ts:
```diff
@@ -0,0 +1,183 @@
+    frame = { shape: "rectangle" }
+  }
+  if (frame.shape === undefined || frame.shape === "none") return false;
+  const frameCurve = computeFrame({ frame: frame.shape, range: layout.range, transform });
```
I would pass this in to avoid duplicate calls to `computeFrame`. it would also remove a dependency to another file which could be good.

claudiaareneee commented on core/backend/src/annotations/LeaderGeometry.ts:
```diff
@@ -0,0 +1,183 @@
+ * @returns `true` if at least one leader with a terminator was successfully appended; otherwise, `false`.
+ * @beta
+ */
+export function appendLeadersToBuilder(builder: ElementGeometry.Builder, leaders: TextAnnotationLeader[], layout: TextBlockLayout, transform: Transform, params: GeometryParams, frame?: TextFrameStyleProps,): boolean {
+  // If there is no frame, use a rectangular frame to compute the attachmentPoints for leaders.
```
```suggestion
export function appendLeadersToBuilder(builder: ElementGeometry.Builder, leaders: TextAnnotationLeader[], layout: TextBlockLayout, transform: Transform, params: GeometryParams, frame?: TextFrameStyleProps): boolean {
  let result = true;

  // If there is no frame, use a rectangular frame to compute the attachmentPoints for leaders.
```

claudiaareneee commented on core/backend/src/annotations/LeaderGeometry.ts:
```diff
@@ -0,0 +1,183 @@
+
+    leaderLinePoints.push(attachmentPoint)
+
+    builder.appendGeometryQuery(LineString3d.create(leaderLinePoints));
```
```suggestion
    result = result && builder.appendGeometryQuery(LineString3d.create(leaderLinePoints));
```

claudiaareneee commented on core/backend/src/annotations/LeaderGeometry.ts:
```diff
@@ -0,0 +1,183 @@
+    const termPointB = basePoint.plusScaled(termY.negate(), terminatorHeight);
+    builder.appendGeometryQuery(LineString3d.create([termPointA, leader.startPoint, termPointB]));
+
+    return true;
```
```suggestion
    return result;
```

claudiaareneee reviewed (commented):
I left some suggestions. Also I have a feeling that tests are going to be asked for

SheetalNair commented on core/common/src/annotation/TextStyle.ts:
```diff
@@ -204,7 +253,9 @@ export class TextStyleSettings {
-      && this.superScriptOffsetFactor === other.superScriptOffsetFactor && this.superScriptScale === other.superScriptScale;
+      && this.superScriptOffsetFactor === other.superScriptOffsetFactor && this.superScriptScale === other.superScriptScale && this.leaderColor === other.leaderColor
+      && this.wantElbow === other.wantElbow && this.elbowLength === other.elbowLength
+      && this.terminatorHeight === other.terminatorHeight && this.terminatorWidth === other.terminatorWidth;
```
done

pmconne commented on common/api/core-common.api.md:
```diff
@@ -10105,13 +10147,15 @@ export class TextStyleSettings {
     readonly color: TextStyleColor;
     static defaultProps: Readonly<Required<TextStyleSettingsProps>>;
     static defaults: TextStyleSettings;
+    readonly elbowLength: number;
```
I think it would be helpful to define an interface type that contains all of the leader-related settings, and give TextStyleSettings a single property of that type. Mixing them all in as top-level properties makes it very difficult to tell which are related to leaders, difficult to write APIs that operate on the leader-specific properties as a unit, and make conflicts/confusion likely in future (for example, if we add new annotations for which elbows are relevant).

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -65,6 +65,45 @@ export interface TextFrameStyleProps {
+ * The leader line can be attached to a key point, to the first or last line of text, or to the nearest point on the text block frame.
+ * @beta
+ */
+export type LeaderAttachmentMode =
```
```suggestion
export type LeaderAttachment =
```

This type **has** a `mode`, it is not itself a mode.

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -65,6 +65,45 @@ export interface TextFrameStyleProps {
+  | { mode: "Nearest" };
+
+/**
+ * Specifies the possible positions for the leader attachment point relative to the annotation when attachmentMode is TextPoint.
```
I assume the leader attaches to the annotation's frame. How does this work when the frame is not rectangular? e.g., what does `TopLeft` mean if the frame is a triangle?

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -235,16 +294,40 @@ export class TextAnnotation {
       && this.frame?.border === other.frame?.border
       && this.frame?.borderWeight === other.frame?.borderWeight;
 
+    const leadersMatch = this.areLeadersEqual(this.leaders, other.leaders);
```
Do the fast and easy comparison first, returning early if they don't match. Then only compare frames if necessary, and finally only compare leaders if necessary.

pmconne reviewed (changes requested):
You don't appear to have any test coverage for the code in LeaderGeometry.ts.

SheetalNair commented on common/api/core-common.api.md:
```diff
@@ -9881,10 +9898,35 @@ export type TextAnnotationFillColor = TextStyleColor | "background";
+export interface TextAnnotationLeaderProps {
+    // (undocumented)
+    attachmentMode: LeaderAttachmentMode;
+    // (undocumented)
```
done

SheetalNair commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -65,6 +65,45 @@ export interface TextFrameStyleProps {
 
+/**
+ * Specifies where to attach a leader line to a [[TextAnnotation]].
+ * The leader line can be attached to a key point, to the first or last line of text, or to the nearest point on the text block frame.
```
updated documentation

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -69,6 +69,69 @@ export interface TextFrameStyleProps {
 }
 
+/**
+ * Describes how to attach a leader to a frame around [[textBlock]].
```
```suggestion
 * Describes how to attach a [[TextAnnotationLeader]] to the frame around a [[TextBlock]].
```

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -69,6 +69,69 @@ export interface TextFrameStyleProps {
+   * If not specified, the leader will be a straight line from startPoint to the point on the frame.
+   * For now, intermediate points are a set of points which create additional LineSegments in the leader, but there could be intermediate shapes instead of straight LineSegments in future*/
+  intermediatePoints?: Point3d[];
+  /** Optional style overrides for the leader. If not specified, the leader will use the default style from [[TextStyleSettings]].
```
In LeaderGeometry.ts, you seem only to look at `styleOverrides`, not falling back to `TextStyle.leader`.

SheetalNair replied:
This is the PR I was referring to https://github.com/iTwin/itwinjs-core/pull/8313 . It introduces a slightly different way to store and fetch text styles, so it would better to have those changes there.

Josh-Schifter replied:
Because, until that PR, none of the text properties are coming from the style so we chose to address all the properties together in one PR https://github.com/iTwin/itwinjs-core/pull/8313 which has been developed in parallel to this one.

SheetalNair commented:
> You don't appear to have any test coverage for the code in LeaderGeometry.ts.

Added tests for LeaderGeometry.  I have introduced a AnnotationTestUtils.ts file, as there was some code from other tests that needed to be reused in my tests. please let me know if that's fine.

pmconne commented:
The geometry tests don't exercise any of the `styleOverrides` except for `wantElbow`. Verify the geometry produced with different elbow lengths and terminator width/height factors, different attachment modes, etc.

SheetalNair commented on core/backend/src/test/annotations/LeaderGeometry.test.ts:
```diff
@@ -0,0 +1,192 @@
+      ]
+      const result = appendLeadersToBuilder(builder, leaders, layout, transform, defaultParams, frame);
+      expect(result).to.be.true;
+      expect(builder.geometries.length).to.be.greaterThan(0);
```
Added more details to tests
